README.MD

#The pydeflate Package



.. image:: https://img.shields.io/pypi/v/pydeflate.svg
:target: https://pypi.python.org/pypi/pydeflate

.. image:: https://readthedocs.org/projects/pydeflate/badge/?version=latest
:target: https://pydeflate.readthedocs.io/en/latest/?version=latest
:alt: Documentation Status

.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
:target: https://github.com/psf/black
:alt: Code style: black

.. image:: https://pepy.tech/badge/pydeflate/month
:target: https://pepy.tech/project/pydeflate
:alt: Downloads


Pydeflate is a Python package to convert flows data to constant prices. This can be done
from any source currency to any desired base year and currency. Pydeflate can also
be used to convert constant data to current prices and to convert from one currency
to another (in current and constant prices). Users can choose the source of the
exchange and deflator/prices data (IMF, World Bank or OECD DAC).


* Free software: MIT license
* Documentation: https://pydeflate.readthedocs.io.


Installation
------------

pydeflate is registered at PyPI. From the command line:

    pip install pydeflate --upgrade

Alternatively, the source code is available on [GitHub](https://github.com/jm-rivera/pydeflate).


## Usage


### Basic usage


Convert data expressed in current USD prices to constant EUR prices for a given base year:

    import pydeflate
    import pandas as pd
    
    #example data
    data = {'iso_code': ['FRA','USA', 'GTM'],
            'year': [2017, 2017, 2017],
            'value': [50, 100, 200]}
    
    #create an example dataframe, in current USD prices
    df = pd.DataFrame.from_dict(data)
    
    #convert to EUR 2015 constant prices
    df_constant = pydeflate.deflate(
        df = df,
        base_year = 2015,
        source = 'wb', #exchange/deflators from the IMF
        method = 'gdp',
        source_currency = "USA", #since data is in USD
        target_currency = "EMU", #we want the result in constant EUR
        iso_column = "iso_code",
        date_column = "year",
        source_col = "value",
        target_col = "value_constant",
    ) 
            
    print(df_constant)

This results in a dataframe containing a new column `value_constant` in 2015 constant
prices. In the background, pydeflate takes into account:

- changes in princes, through a gdp deflator in this case
- changes in exchange rates overtime

Pydeflate can also handle data that is expressed in local currency units. In that
case, users can specify `LCU` as the source currency.

    import pydeflate
    import pandas as pd
    
    #example data
    data = {'iso_code': ['GBR','GBR', 'JPN'],
            'date': [2011, 2015, 2015],
            'value': [100, 100, 100]}
    
    #create an example dataframe, in current local currency units 
    df = pd.DataFrame.from_dict(data)
    
    #convert to USD 2018 constant prices
    df_constant = pydeflate.deflate(
        df = df,
        base_year = 2018,
        source = 'imf', #exchange/deflators from the World Bank
        method = 'pcpi',
        source_currency = "LCU", #since data is in LCU
        target_currency = "USA", #to get data in USD
        iso_column = "iso_code",
        date_column = "date",
        source_col = "value",
        target_col = "value", #to not create a new column
    ) 
            
    print(df_constant)

Users can also convert a dataset expressed in constant prices to current prices
using pydeflate. To avoid introducing errors, users should know which methodology/
data was used to create constant prices by the original source. The basic usage is
the same as before, but the `reverse` parameter is set to `True`.

For example, to convert DAC data expressed in 2016 USD constant prices to current
US dollars:

    import pydeflate
    import pandas as pd
    
    #example data
    data = {'iso_code': ['USA','ITA', 'ITA'],
            'date': [2010, 2016, 2018],
            'value': [100, 100, 100]}
    
    #create an example dataframe, in current local currency units 
    df = pd.DataFrame.from_dict(data)
    
    #convert to USD 2018 constant prices
    df_current = pydeflate.deflate(
        df = df,
        base_year = 2016,
        source = 'oecd_dac', 
        source_currency = "USA", #since data is in USD constant
        target_currency = "USA", #to get data in USD
        iso_column = "iso_code",
        date_column = "date",
        source_col = "value",
        target_col = "value_current", 
        reverse = True,   
    ) 
            
    print(df_current)


### Data source and method options


A `source` and a `method` for the exchange and price/gdp deflators must be chosen.
The appropriate combination depends on the objectives of the project or the nature
of the original data.

In terms of price or GDP deflators, pydeflate provides the following `methods`:

- World Bank ("wb"):

    - `gdp`: in order to use GDP deflators.
    - `gdp_linked`: to use the World Bank's GDP deflator series which has been linked
      to produce a consistent time series to counteract breaks in series over time
      due to changes in base years, sources or methodologies.
    - 'cpi': to use Consumer Price Index data

- International Monetary Fund World Economic Outlook ("imf"):

    - `pcpi`: in order to use Consumer Price Index data.
    - `pcpie`: to use end-of-period Consumer Price Index data (e.g for December each
      year).

- OECD Development Assistance Committee ("oecd_dac"):

    - None: for consistency with how the DAC calculates deflators, only their methodology
      is accepted/used with this data.

The source of the exchange rate data depends on the source selected. Both "imf"
and "wb" use data from the International Monetary Fund (`LCU per US$, yearly average`).
The OECD Development Assistance Committee data uses different exchange rates. When
`oecd_dac` is selected as the source, the OECD DAC exchange rates (`LCU per US$`) are
used. Exchange rates between two non USD currency pairs are derived from the LCU to USD
exchange rates selected.


### Additional features


Pydeflate relies on data from the World Bank, IMF and OECD for its calculations. This
data is updated periodically. If the version of the data stored in the user's computer
is older than 50 days, pydeflate will show a warning on import.

Users can always update the underlying data by using:

    import pydeflate
    
    pydeflate.update_all_data()


Pydeflate also provides users with a tool to exchange figures from one currency to
another, without applying any deflators. This should only be used on numbers expressed
in current prices, however.

In this version of pydeflate, the dataframe must contain a column with iso3 country
codes called `iso_code`.

For example, to convert numbers in current Local Currency Units (LCU) to current
Canadian Dollars:

    import pydeflate
    import pandas as pd
    
    #example data
    data = {'iso_code': ['GBR','CAN', 'JPN'],
            'date': [2011, 2015, 2015],
            'value': [100, 100, 100]}
    
    #create an example dataframe, in current local currency units 
    df = pd.DataFrame.from_dict(data)
    
    #convert to USD 2018 constant prices
    df_can = pydeflate.exchange(
        df = df,
        source_currency = "LCU", #since data is in LCU
        target_currency = "CAN", #to get data in Canadian Dollars
        rates_source = 'wb', #this is the same as IMF exchange rates
        value_column = 'value',
        target_column = 'value_CAN',
        date_column = "date",
    ) 
            
    print(df_can)


## Credits

This package relies on data from the following sources:

- OECD DAC: https://www.oecd.org/dac/ (Official Development assistance data (DAC1),
  DAC deflators, and exchange rates used by the DAC)
- IMF World Economic Outlook: https://www.imf.org/en/Publications/WEO (GDP and
  price deflators)
- World Bank DataBank: https://databank.worldbank.org/home.aspx (exchange rates,
  GDP and price deflators)

This data is provided based on the terms and conditions set by the original sources

HISTORY.md
# History

## 0.1.4 (2021-11-22)

* First release on PyPI.


pydeflate/utils.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import datetime
import json
import warnings

import numpy as np
import pandas as pd

from pydeflate import config


def _diff_from_today(date: datetime.datetime):
    """Compare to today"""

    today = datetime.datetime.today()

    return (today - date).days


def warn_updates():
    with open(config.paths.data + r"/data_updates.json") as file:
        updates = json.load(file)

    for source, date in updates.items():
        d = datetime.datetime.strptime(date, "%Y-%m-%d")
        if _diff_from_today(d) > 50:
            message = (
                f'\n\nThe underlying data for "{source}" has not been updated'
                f" in over {_diff_from_today(d)} days. \nIn order to use"
                " pydeflate with the most recent data, please run:\n"
                '"pydeflate.update_all_data()"\n\n'
            )
            warnings.warn(message)


def update_update_date(source: str):
    """Update the most recent update date for data to today"""

    today = datetime.datetime.today().strftime("%Y-%m-%d")

    with open(config.paths.data + r"/data_updates.json") as file:
        updates = json.load(file)

    updates[source] = today

    with open(config.paths.data + r"/data_updates.json", "w") as outfile:
        json.dump(updates, outfile)


oecd_codes = {
    1: "AUT",
    2: "BEL",
    3: "DNK",
    4: "FRA",
    5: "DEU",
    6: "ITA",
    7: "NLD",
    8: "NOR",
    9: "PRT",
    10: "SWE",
    11: "CHE",
    12: "GBR",
    18: "FIN",
    20: "ISL",
    21: "IRL",
    22: "LUX",
    40: "GRC",
    50: "ESP",
    61: "SVN",
    68: "CZE",
    69: "SVK",
    75: "HUN",
    76: "POL",
    301: "CAN",
    302: "USA",
    701: "JPN",
    742: "KOR",
    801: "AUS",
    820: "NZL",
    918: "EUI",
    30: "CYP",
    45: "MLT",
    55: "TUR",
    62: "HRV",
    70: "LIE",
    72: "BGR",
    77: "ROU",
    82: "EST",
    83: "LVA",
    84: "LTU",
    87: "RUS",
    130: "DZA",
    133: "LBY",
    358: "MEX",
    543: "IRQ",
    546: "ISR",
    552: "KWT",
    561: "QAT",
    566: "SAU",
    576: "ARE",
    611: "AZE",
    613: "KAZ",
    732: "TWN",
    764: "THA",
    765: "TLS",
}


emu = [
    "AUT",
    "BEL",
    "CYP",
    "EST",
    "FIN",
    "FRA",
    "DEU",
    "GRC",
    "IRL",
    "ITA",
    "LVA",
    "LTU",
    "LUX",
    "MLT",
    "NLD",
    "PRT",
    "SVK",
    "SVN",
    "ESP",
]


def clean_number(number):
    """Clean a number and return as float"""
    import re

    if not isinstance(number, str):
        number = str(number)

    number = re.sub(r"[^0-9.]", "", number)

    if number == "":
        return np.nan

    return float(number)


def base_year(df: pd.DataFrame, date_col: str = "date") -> dict:

    """Return dictionary of base years by iso_code"""

    return df.loc[df.value == 100].set_index("iso_code")[date_col].to_dict()


def value_index(df: pd.DataFrame, base_dict: dict) -> pd.Series:
    """Reindex a series given a base year dictionary"""

    data = df.copy()
    df_ = df.copy()

    df_["base"] = df_.iso_code.map(base_dict)
    df_ = df_.loc[df_.year == df_.base]

    base_values = (
        df_.loc[lambda d: d.value.notna()]
        .round(5)
        .set_index("iso_code")["value"]
        .to_dict()
    )

    return round(100 * data.value / data.iso_code.map(base_values), 3)


def rebase(df_: pd.DataFrame, base_year: int) -> pd.Series:
    """Rebase values to a given base year"""

    base_values = (
        df_.loc[df_.year.dt.year == base_year].set_index("iso_code")["value"].to_dict()
    )

    return round(100 * df_.value / df_.iso_code.map(base_values), 3)


def check_method(method: str, methods: dict):
    """Check whether a given method is in a methods dictionary"""

    if method not in methods.keys():
        raise ValueError(f'Method "{method}" not valid. Please see documentation.')


def check_year_as_number(df: pd.DataFrame, date_column: str) -> (pd.DataFrame, bool):
    """Check whether the date column contains an int instead of datetime.
    This changes the column to datetime and returns a flag"""

    if pd.api.types.is_numeric_dtype(df[date_column]):
        df[date_column] = pd.to_datetime(df[date_column], format="%Y")
        year_as_number = True
    else:
        year_as_number = False

    return df, year_as_number
